arangod/Cluster/RebootTracker.cpp

```diff
@@ -66,296 +95,117 @@ RebootTracker::RebootTracker(RebootTracker::SchedulerPointer scheduler)
-  for_each_iter(_rebootIds.begin(), _rebootIds.end(), [&](auto const curIt) {
-    auto const& serverId = curIt->first;
-    auto& oldRebootId = curIt->second;
-    auto const& newIt = state.find(serverId);
-
-    if (newIt == state.end()) {
-      // Try to schedule all callbacks for serverId.
-      // If that didn't throw, erase the entry.
-      LOG_TOPIC("88858", INFO, Logger::CLUSTER)
-          << "Server " << serverId << " removed, aborting its old jobs now.";
-      scheduleAllCallbacksFor(serverId);
-      auto it = _callbacks.find(serverId);
-      if (it != _callbacks.end()) {
-        TRI_ASSERT(it->second.empty());
-        _callbacks.erase(it);
-      }
-      _rebootIds.erase(curIt);
-    } else {
-      TRI_ASSERT(serverId == newIt->first);
-      auto const& newRebootId = newIt->second;
-      TRI_ASSERT(oldRebootId <= newRebootId);
-      if (oldRebootId < newRebootId) {
-        LOG_TOPIC("88857", INFO, Logger::CLUSTER)
-            << "Server " << serverId
-            << " gone or rebooted, aborting its old jobs now.";
-        // Try to schedule all callbacks for serverId older than newRebootId.
-        // If that didn't throw, erase the entry.
-        scheduleCallbacksFor(serverId, newRebootId);
-        oldRebootId = newRebootId;
-      }
-    }
-  });
-
-  // Look whether there are servers that are still unknown
-  // (note: we could shortcut this and return if the sizes are equal, as at
-  // this point, all entries in _rebootIds are also in state)
-  for (auto const& newIt : state) {
-    auto const& serverId = newIt.first;
-    auto const& rebootId = newIt.second;
-    auto const rv = _rebootIds.try_emplace(serverId, rebootId);
-    auto const inserted = rv.second;
-    // If we inserted a new server, we may NOT already have any callbacks for
-    // it!
-    TRI_ASSERT(!inserted || _callbacks.find(serverId) == _callbacks.end());
   }
+  _state = std::move(state);
 }
 
-CallbackGuard RebootTracker::callMeOnChange(
-    RebootTracker::PeerState const& peerState, RebootTracker::Callback callback,
-    std::string callbackDescription) {
-  MUTEX_LOCKER(guard, _mutex);
-
-  auto const rebootIdIt = _rebootIds.find(peerState.serverId());
-
-  // We MUST NOT insert something in _callbacks[serverId] unless
-  // _rebootIds[serverId] exists!
-  if (rebootIdIt == _rebootIds.end()) {
-    std::string const error = [&]() {
-      std::stringstream strstream;
-      strstream << "When trying to register callback '" << callbackDescription
-                << "': "
-                << "The server " << peerState.serverId() << " is not known. "
-                << "If this server joined the cluster in the last seconds, "
-                   "this can happen.";
-      return strstream.str();
-    }();
+CallbackGuard RebootTracker::callMeOnChange(RebootTracker::PeerState peer,
+                                            Callback callback,
+                                            std::string_view description) {
```

Comment:
I don't care much about whether the description is passed as a `string_view` or a `string`, but I have to ask: Why did you change that? The way I see it, we now always create a copy (when the `DescriptedCallback` is constructed). Previously, the caller had the option to move, which could optionally avoid it.

Reply:
Ok, https://github.com/arangodb/arangodb/pull/16375/commits/5107c2495127e272174740e3cba5383327ad6072

> we now always create a copy

No, if this Peer outdated then not

> Previously, the caller had the option to move

Yes, but we never use this opportunity 


I think we have a problem in that ServerId is string (not in this code, but in general):
We almost always create copy of it: from vpacks although we store that vpacks, to other functions, etc 

Btw maybe you know why `ServerID` is `std::string`, why isn't `uint64_t`?

Reply:
> No, if this Peer outdated then not

That case virtually never happens and is definitely not the case to optimize for.

> Yes, but we never use this opportunity

And now we won't in the future, either. :) Also, it's not true: In half the cases where this is called, there's a string created anyway, which is moved here. In those cases, we now make an additional copy. In the other cases, the copy is now made here instead of at the caller's site.

> I think we have a problem in that ServerId is string (not in this code, but in general):
> We almost always create copy of it: from vpacks although we store that vpacks, to other functions, etc
>
>Btw maybe you know why ServerID is std::string, why isn't uint64_t?

I don't know how that evolved historically. I'd also like to have a `uint64_t` or something else of fixed length (could even be a string), at least something we can pass around without allocating. Could probably be done by having a map somewhere, that can be consulted during (de)serialization. As a simple idea, maybe the UUIDs could be converted to an 128bit int, though that would forget the server type prefix (i.e. `PRMR-`, `COOR-`, `AGNT-`); I don't know the impact of that without more research. And server short names can't be forgotten either. So I guess that it could be done.

Reply:
Ok, anyway done?

Reply:
Maybe you didn't see, but I reverted to string

Reply:
string_view better for description because we pass string literal, so in that case we made copy always if and when needed

Reply:
But we don't always pass a string literal:
```cpp
        engine->rebootTrackers().emplace_back(ci.rebootTracker().callMeOnChange(
            cluster::RebootTracker::PeerState(server, rebootId), std::move(f),
            std::move(comment)));
```
(from ExecutionEngine.cpp)

```cpp
      auto rGuard =
          std::make_unique<RebootCookie>(ci.rebootTracker().callMeOnChange(
              RebootTracker::PeerState(serverId, rebootId), std::move(f),
              std::move(comment)));
```
(from RestReplicationHandler.cpp)

Both are dynamically calculated `std::string`s. These now always copy.

And for the calls with literals, it does not make a difference: Either the copy is made at the caller's site (when passed as a std::string), or when constructing the DescriptedCallback (when passed as a std::string_view). But a copy is made either way.

So as far as I can see, this is a pessimization.

Reply:
> But if we pass always const ref string, or literal string_view always better than other possibilities, them not the same

Yes, agreed.